File: app/src/main/java/com/njamb/geodrink/authentication/LoginActivity.java
package com.njamb.geodrink.authentication;

import android.app.Activity;
import android.app.ProgressDialog;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.View;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.mobsandgeeks.saripaar.ValidationError;
import com.mobsandgeeks.saripaar.Validator;
import com.mobsandgeeks.saripaar.annotation.Email;
import com.mobsandgeeks.saripaar.annotation.NotEmpty;
import com.mobsandgeeks.saripaar.annotation.Password;
import com.njamb.geodrink.R;

import java.util.List;

public class LoginActivity extends AppCompatActivity implements Validator.ValidationListener {
    private static final String TAG = "LoginActivity";
    private static final int REQUEST_REGISTER = 1;

    private Validator mValidator = null;
    @NotEmpty
    @Email private EditText email;
    @Password private EditText password;

    private ProgressDialog pd;
    private FirebaseAuth mAuth;


    @Override
    public void onBackPressed() {
        finishAffinity();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_login);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        getSupportActionBar().setDisplayHomeAsUpEnabled(false);

        mValidator = new Validator(this);
        mValidator.setValidationListener(this);

        mAuth = FirebaseAuth.getInstance();
        if (mAuth.getCurrentUser() != null) finish(); // TODO: how to remove this hack?

        email = (EditText) findViewById(R.id.login_et_username);
        password = (EditText) findViewById(R.id.login_et_password);

        Button login = (Button) findViewById(R.id.login_btn_login);
        login.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mValidator.validate();
            }
        });

        Button register = (Button) findViewById(R.id.login_btn_register);
        register.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent registerActivity = new Intent(LoginActivity.this, RegisterActivity.class);
                startActivityForResult(registerActivity, REQUEST_REGISTER);
            }
        });

        Button btnResetPass = (Button) findViewById(R.id.btn_rst_pass);
        btnResetPass.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(LoginActivity.this, ResetPasswordActivity.class));
            }
        });

        configProgressDialog();
    }

    private void configProgressDialog() {
        pd = new ProgressDialog(this, R.style.TransparentProgressDialogStyle);
        pd.setProgressStyle(android.R.style.Widget_ProgressBar_Small);
        pd.setIndeterminate(true);
        pd.setCancelable(false);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        switch (requestCode) {
            case REQUEST_REGISTER:
                if (resultCode == Activity.RESULT_OK) {
                    Toast.makeText(this, "Successful registration!", Toast.LENGTH_SHORT).show();
                }
        }
    }

    private void loginUser() {
        String email = this.email.getText().toString();
        String pass = password.getText().toString();

        getWindow().addFlags(WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE);
        pd.show();
        mAuth.signInWithEmailAndPassword(email, pass)
                .addOnCompleteListener(this, new OnCompleteListener<AuthResult>() {
                    @Override
                    public void onComplete(@NonNull Task<AuthResult> task) {
                        getWindow().clearFlags(WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE);
                        pd.dismiss();

                        if (task.isSuccessful()) {
                            Log.d(TAG, "signInWithEmail:success");
                            finish();
                        } else {
                            Log.w(TAG, "signInWithEmail:failure", task.getException());
                            Toast.makeText(LoginActivity.this,
                                    R.string.msg_auth_fail,
                                    Toast.LENGTH_LONG).show();
                        }
                    }
                });
    }

    @Override
    public void onValidationSucceeded() {
        loginUser();
    }

    @Override
    public void onValidationFailed(List<ValidationError> errors) {
        for (ValidationError error : errors) {
            View view = error.getView();
            String message = error.getCollatedErrorMessage(this);
            if (view instanceof EditText) {
                ((EditText) view).setError(message);
            }
            else {
                Toast.makeText(this, message, Toast.LENGTH_SHORT).show();
            }
        }
    }
}

File: app/src/main/java/com/njamb/geodrink/authentication/LoginActivity.kt
package com.njamb.geodrink.authentication

import android.app.Activity
import android.app.ProgressDialog
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.util.Log
import android.view.View
import android.view.WindowManager
import android.widget.Button
import android.widget.EditText
import android.widget.Toast

import com.google.android.gms.tasks.OnCompleteListener
import com.google.android.gms.tasks.Task
import com.google.firebase.auth.AuthResult
import com.google.firebase.auth.FirebaseAuth
import com.mobsandgeeks.saripaar.ValidationError
import com.mobsandgeeks.saripaar.Validator
import com.mobsandgeeks.saripaar.annotation.Email
import com.mobsandgeeks.saripaar.annotation.NotEmpty
import com.mobsandgeeks.saripaar.annotation.Password
import com.njamb.geodrink.R

class LoginActivity : AppCompatActivity(), Validator.ValidationListener {

    private lateinit var mValidator: Validator
    @NotEmpty
    @Email
    private lateinit var email: EditText
    @Password private lateinit var password: EditText

    private lateinit var pd: ProgressDialog
    private lateinit var mAuth: FirebaseAuth


    override fun onBackPressed() {
        finishAffinity()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_login)
        val toolbar = findViewById(R.id.toolbar) as Toolbar
        setSupportActionBar(toolbar)

        supportActionBar!!.setDisplayHomeAsUpEnabled(false)

        mValidator = Validator(this)
        mValidator.setValidationListener(this)

        mAuth = FirebaseAuth.getInstance()
        if (mAuth.currentUser != null) finish() // TODO: how to remove this hack?

        email = findViewById(R.id.login_et_username) as EditText
        password = findViewById(R.id.login_et_password) as EditText

        val login = findViewById(R.id.login_btn_login) as Button
        login.setOnClickListener { mValidator.validate() }

        val register = findViewById(R.id.login_btn_register) as Button
        register.setOnClickListener {
            val registerActivity = Intent(this@LoginActivity, RegisterActivity::class.java)
            startActivityForResult(registerActivity, REQUEST_REGISTER)
        }

        val btnResetPass = findViewById(R.id.btn_rst_pass) as Button
        btnResetPass.setOnClickListener { startActivity(Intent(this@LoginActivity, ResetPasswordActivity::class.java)) }

        configProgressDialog()
    }

    private fun configProgressDialog() {
        pd = ProgressDialog(this, R.style.TransparentProgressDialogStyle)
        pd.setProgressStyle(android.R.style.Widget_ProgressBar_Small)
        pd.isIndeterminate = true
        pd.setCancelable(false)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        super.onActivityResult(requestCode, resultCode, data)

        when (requestCode) {
            REQUEST_REGISTER -> if (resultCode == Activity.RESULT_OK) {
                Toast.makeText(this, "Successful registration!", Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun loginUser() {
        val email = this.email.text.toString()
        val pass = password.text.toString()

        window.addFlags(WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE)
        pd.show()
        mAuth.signInWithEmailAndPassword(email, pass)
                .addOnCompleteListener(this) { task ->
                    window.clearFlags(WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE)
                    pd.dismiss()

                    if (task.isSuccessful) {
                        Log.d(TAG, "signInWithEmail:success")
                        finish()
                    } else {
                        Log.w(TAG, "signInWithEmail:failure", task.exception)
                        Toast.makeText(this@LoginActivity,
                                       R.string.msg_auth_fail,
                                       Toast.LENGTH_LONG).show()
                    }
                }
    }

    override fun onValidationSucceeded() {
        loginUser()
    }

    override fun onValidationFailed(errors: List<ValidationError>) {
        for (error in errors) {
            val view = error.view
            val message = error.getCollatedErrorMessage(this)
            if (view is EditText) {
                view.error = message
            } else {
                Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
            }
        }
    }

    companion object {
        private val TAG = "LoginActivity"
        private val REQUEST_REGISTER = 1
    }
}
